File: Instructors Area/Instructor-Materials/Spark-Essentials/Labs/RDD-Fundamentals/Philadelphia-Crime/Scala/RDD-Solution.scala
// Databricks notebook source exported at Wed, 4 Nov 2015 01:36:21 UTC
// MAGIC %md # Solutions for RDD Lab (Scala)

// COMMAND ----------

// MAGIC %md
// MAGIC ## Question 1
// MAGIC 
// MAGIC **Does the RDD _actually_ contain the data right now?**
// MAGIC 
// MAGIC No. RDDs are _lazy_.

// COMMAND ----------

// MAGIC %md 
// MAGIC ## Question 2
// MAGIC 
// MAGIC **Why can't you just do `baseRDD.drop(1)`?**
// MAGIC 
// MAGIC 1. There _is_ no such methods on RDDs, unlike Scala collections.
// MAGIC 2. Even if there were, RDDs are _partitioned_, and a `drop()` call would run on every node, dropping the first element of each partition. On one partition, that element would be the header, but not on the other partitions.

// COMMAND ----------

// MAGIC %md ## Exercise 1

// COMMAND ----------

// TAKE NOTE: We are deliberately only some of the fields we need for
// this lab. There's no sense dragging around more data than we need.
case class CrimeData(dateString: String,
                     timeString: String,
                     offense: String,
                     latitude: String,
                     longitude: String)
                     
val dataRDD = noHeaderRDD.map { line =>
  val cols = line.split(",")
  CrimeData(dateString = cols(10), // DISPATCH_DATE
            timeString = cols(11), // DISPATCH_TIME
            offense    = cols(6),  // TEXT_GENERAL_CODE
            latitude   = cols(7),  // POINT_X
            longitude  = cols(8))  // POINT_Y
}
dataRDD.take(10).foreach(println)

// COMMAND ----------

// MAGIC %md ## Exercise 2

// COMMAND ----------

val groupedByOffenseRDD = dataRDD.groupBy { data => data.offense }

// What does this return? You'll need to know for the next step.
groupedByOffenseRDD.take(1).foreach(println)

// COMMAND ----------

// MAGIC %md ## Exercise 3

// COMMAND ----------

// Here's one way to do it:

val offenseCounts = groupedByOffenseRDD2.map { case (offense, items) => (offense, items.size) }.collect()
for ((offense, count) <- offenseCounts) {
  println(s"$offense: $count")
}

// COMMAND ----------

// But here's a better way. Note the use of Scala's "f" string interpolator,
// which is kind of like printf, but with compile-time type safety. For
// more information on the "f" interpolator, see
// http://docs.scala-lang.org/overviews/core/string-interpolation.html

val offenseCounts = dataRDD.map(item => (item.offense, item)).countByKey()
for ((offense, count) <- offenseCounts) {
  println(f"$offense%30s $count%5d")
}

// COMMAND ----------

// MAGIC %md ## Exercise 4

// COMMAND ----------

println(baseRDD.partitions.length)
println(groupedByOffenseRDD2.partitions.length)

// COMMAND ----------

// MAGIC %md ## Exercise 5

// COMMAND ----------

// It's okay to cache the base RDD. It's not very large. Proof:
println(s"Count of lines: ${baseRDD.count}")
val totalChars = baseRDD.map(_.length).reduce(_ + _)
println(s"Count of characters (Unicode): ${totalChars}")

// They're all about the same size, since we didn't filter any data out. However,
// since we're mostly working with the `cleanedRDD`, that's the one to cache.
cleanedRDD.cache()

// COMMAND ----------

// MAGIC %md ## Exercise 6

// COMMAND ----------

val resultRDD1 = cleanedRDD.filter(_.offense.toLowerCase contains "homicide").map { item => (item.offense, 1) }.reduceByKey(_ + _)

for ((method, count) <- resultRDD1.collect())
  println(f"$method%10s $count%d")

File: Spark-Essentials-TE2/Spark-Essentials/Labs/RDD-Fundamentals/Philadelphia-Crime/Python/RDD-Solution.py
# Databricks notebook source exported at Wed, 4 Nov 2015 01:33:16 UTC
# MAGIC %md
# MAGIC # Solutions for RDD Lab (Python)

# COMMAND ----------

# MAGIC %md ## Exercise 1

# COMMAND ----------

CrimeData = namedtuple('CrimeData', ['date_string', 'time_string', 'offense', 'latitude', 'longitude'])

def map_line(line):
  cols = line.split(",")
  return CrimeData(date_string=cols[10], time_string=cols[11], offense=cols[6], latitude=cols[7], longitude=cols[8])
  
data_rdd = no_header_rdd.map(map_line)
print data_rdd.take(10)

# COMMAND ----------

# MAGIC %md ## Exercise 2

# COMMAND ----------

grouped_by_offense_rdd = data_rdd.groupBy(lambda data: data.offense)

# What does this return? You'll need to know for the next step.
print grouped_by_offense_rdd.take(10)


# COMMAND ----------

# MAGIC %md ## Exercise 3

# COMMAND ----------

# Here's one way to do it:

offense_counts = grouped_by_offense_rdd2.map(lambda g: (g[0], len(g[1]))).collect()
for offense, count in offense_counts:
  print "{0:30s} {1:d}".format(offense, count)

# COMMAND ----------

# But here's a better way:

offense_counts = cleaned_rdd.map(lambda item: (item.offense, item)).countByKey()
for offense, counts in offense_counts.items():
  print "{0:30s} {1:d}".format(offense, counts)

# COMMAND ----------

# MAGIC %md ## Exercise 4

# COMMAND ----------

print base_rdd.getNumPartitions()
print grouped_by_offense_rdd2.getNumPartitions()

# COMMAND ----------

# MAGIC %md ## Exercise 5

# COMMAND ----------

# It's okay to cache the base RDD. It's not very large. Proof:
print "Count of lines: {0}".format(base_rdd.count())
totalChars = base_rdd.map(lambda line: len(line)).reduce(lambda a, b: a + b)
print "Count of characters (Unicode): {0}".format(totalChars)

# They're all about the same size, since we didn't filter any data out. However,
# since we're mostly working with the `cleaned_rdd`, that's the one to cache.
cleaned_rdd.cache()

# COMMAND ----------

# MAGIC %md ## Exercise 6

# COMMAND ----------

result_rdd1 = cleaned_rdd.filter(lambda d: "homicide" in d.offense.lower()).map(lambda d: (d.offense, 1)).reduceByKey(lambda a, b: a + b)

for method, count in result_rdd1.collect():
  print "{0} {1:d}".format(method, count)

# COMMAND ----------

